src/rewrite/steward.ts

```diff
@@ -0,0 +1,331 @@
+    patron = createDefaultPatron(owner, txTimestamp);
+  }
+
+  let wildcard = WildcardNew.load(tokenIdString);
+  if (wildcard == null) {
+    wildcard = createWildcardIfDoesntExist(tokenIdBigInt);
+  }
+
+  let previousTokenOwner = wildcard.owner;
+  let patronOld = PatronNew.load(previousTokenOwner);
+  if (patronOld == null) {
+    patronOld = createNO_OWNERPatron(owner, txTimestamp);
+  }
+
+  // Phase 2: calculate new values.
+
+  // Now even if the patron puts in extra deposit when they buy a new token this will foreclose their old tokens.
+  let heldUntilNewPatron = txTimestamp; //minBigInt(patron.foreclosureTime, txTimestamp); // TODO: use min with foreclosureTime
+  let heldUntilPreviousPatron = txTimestamp; //minBigInt(patron.foreclosureTime, txTimestamp); // TODO: use min with foreclosureTime
+
+  let timeSinceLastUpdatePatron = heldUntilNewPatron.minus(patron.lastUpdated);
+  let timeSinceLastUpdatePreviousPatron = heldUntilPreviousPatron.minus(
+    patronOld.lastUpdated
+  );
+
+  let newPatronTotalTimeHeld =
+    patron.id != "NO_OWNER"
+      ? patron.totalTimeHeld.plus(
+          timeSinceLastUpdatePatron.times(BigInt.fromI32(patron.tokens.length))
+        )
+      : BigInt.fromI32(0);
+  let oldPatronTotalTimeHeld =
+    patronOld.id != "NO_OWNER"
+      ? patronOld.totalTimeHeld.plus(
+          timeSinceLastUpdatePreviousPatron.times(
+            BigInt.fromI32(patronOld.tokens.length)
+          )
+        )
+      : BigInt.fromI32(0);
+
+  let newPatronTokenArray = patron.tokens.concat([wildcard.id]);
+  let itemIndex = patronOld.tokens.indexOf(wildcard.id);
+  let oldPatronTokenArray = patronOld.tokens
+    .slice(0, itemIndex)
+    .concat(patronOld.tokens.slice(itemIndex + 1, patronOld.tokens.length));
+
+  // Phave 3: set+save values.
+
+  patron.lastUpdated = txTimestamp;
+  patron.totalTimeHeld = newPatronTotalTimeHeld;
+  patron.tokens = newPatronTokenArray;
+  patron.save();
+
+  patronOld.lastUpdated = txTimestamp;
+  patronOld.totalTimeHeld = oldPatronTotalTimeHeld;
+  patronOld.tokens = oldPatronTokenArray;
+  patronOld.save();
+
+  wildcard.owner = ownerString;
+  wildcard.save();
+}
+
+export function genericUpdateTimeHeld(
+  owner: Address,
+  txTimestamp: BigInt,
+  steward: Steward,
+  tokenIdBigInt: BigInt
+): void {
+  // PART 1: reading and getting values.
+  let ownerString = owner.toHexString();
+
+  let tokenIdString = tokenIdBigInt.toString();
+
+  let patron = PatronNew.load(ownerString);
+  if (patron == null) {
+    patron = createDefaultPatron(owner, txTimestamp);
+  }
+
+  let wildcard = WildcardNew.load(tokenIdString);
+  if (wildcard == null) {
+    wildcard = createWildcardIfDoesntExist(tokenIdBigInt);
+  }
+
+  // Phase 2: calculate new values.
+
+  // Now even if the patron puts in extra deposit when they buy a new token this will foreclose their old tokens.
+  let heldUntilNewPatron = txTimestamp; //minBigInt(patron.foreclosureTime, txTimestamp); // TODO: use min with foreclosureTime
```

Comment:
Will you update me as to why we are not using the line of commented out code, as your comment says, we should use MIN of foreclosure time and current tx time

Reply:
Big fan of the 
   // PART 1: reading and getting values. 
   // Phase 2: calculate new values. 
   // Phave 3: set+save values. 

Architecture.

Reply:
Because we don't have those values in the 'new' code. :D 

But should be easy to add (for sure!).

The tests will make the architecture shine